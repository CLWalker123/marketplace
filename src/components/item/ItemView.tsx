import React, { useRef } from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { connect } from 'react-redux';
import { FontAwesome } from '@expo/vector-icons';
import { ReduxState } from '../../redux/rootReducer';
import { Item } from './types';
import { generateItemFromItemIdSelector } from './selectors';

interface Props {
  item: Item;
  route: {
    params: NavParams;
  };
}

interface NavParams {
  id: string,
}

export const ItemView = ({ item }: Props) => {
  const timeString = useRef(new Date(Date.now() + item.time_remaining).toTimeString().slice(0,8)).current;
  return (
    <View style={styles.container}>
      <Text style={styles.title}>{item.name}</Text>
      <Text style={styles.price}>{`Â£${item.price / 100}`}</Text>
      <View style={styles.extraInfoContainer}>
        <Text style={styles.timer}>{`Expires at ${timeString}`}</Text>
        <View style={styles.viewsContainer}>
          <FontAwesome name="eye" size={18} style={styles.icon} />
          <Text>{item.views}</Text>
        </View>
      </View>
    </View>
  )
}

const mapStateToProps = (state: ReduxState, props: Props) => {
  return {
    item: generateItemFromItemIdSelector(props.route.params.id)(state),
  }
}

export default connect(mapStateToProps)(ItemView);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingVertical: 12,
    paddingHorizontal: 16,
  },
  extraInfoContainer: {
    justifyContent: 'center',
    alignItems: 'flex-end',
  },
  viewsContainer: {
    flexDirection: 'row',
  },
  title: {
    fontFamily: 'Oxygen_700Bold',
    fontSize: 48,
  },
  price: {
    fontFamily: 'Oxygen_400Regular',
    fontSize: 24,
  },
  timer: {
    fontFamily: 'Oxygen_300Light',
    fontSize: 18,
    color: '#2b2b2b'
  },
  icon: {
    marginHorizontal: 4,
  }
})
