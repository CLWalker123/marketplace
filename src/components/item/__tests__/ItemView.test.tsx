import React from 'react';
import { shallow } from 'enzyme';
import { ItemView } from '../ItemView';

const mockState = {
  item: {
    pending: false,
    error: null,
    items: {},
  }
};

const mockItems = {
  1: {
    id: "1",
    name: "TV",
    price: 400000,
    time_remaining: 18427,
    views: 54,
  },
}

describe('ItemView', () => {
  it('renders the itemiew correctly', () => {
    const wrapper = shallow(
      <ItemView
        item={mockItems["1"]}
        route={{params: {id: '1'}}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});