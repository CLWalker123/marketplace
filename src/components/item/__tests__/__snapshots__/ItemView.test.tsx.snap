// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemView renders the itemiew correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      Object {
        "fontFamily": "Oxygen_700Bold",
        "fontSize": 48,
      }
    }
  >
    TV
  </Text>
  <Text
    style={
      Object {
        "fontFamily": "Oxygen_400Regular",
        "fontSize": 24,
      }
    }
  >
    Â£4000
  </Text>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#2b2b2b",
          "fontFamily": "Oxygen_300Light",
          "fontSize": 18,
        }
      }
    >
      Expires at 00:06:36
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Icon
        allowFontScaling={false}
        name="eye"
        size={18}
        style={
          Object {
            "marginHorizontal": 4,
          }
        }
      />
      <Text>
        54
      </Text>
    </View>
  </View>
</View>
`;
